src/hooks/progress/use-update-tutorial-progress.ts

```diff
@@ -0,0 +1,142 @@
+ * Return two refs in an object, { startRef, endRef },
+ * which can be placed on elements to track tutorial progress.
+ *
+ * On initial load, regardless of startRef or endRef visibility,
+ * if the tutorial does not yet have recorded progress,
+ * the tutorial will be marked as `visited` (0 percent).
+ *
+ * When the startRef element moves out of the viewport,
+ * the tutorial will be marked as `in_progress` (50 percent).
+ *
+ * When the endRef element moves into the viewport,
+ * the tutorial will be marked as `complete` (100 percent).
+ */
+export function useUpdateTutorialProgress({
+	tutorialId,
+	collectionId,
+}: TutorialIdCollectionId): UpdateProgressReturnInterface {
+	// We don't try to update progress unless we're authenticated
+	const { isAuthenticated } = useAuthentication()
+	// We need to know if progress exists, to know whether to "create" or "update"
+	const { tutorialProgressLabel } = useTutorialProgress({
+		tutorialId,
+		collectionId,
+	})
+	// `react-query` mutations are what we'll use to "create" or "update" progress
+	const tutorialProgressMutations = useTutorialProgressMutations()
+	// We don't want to run multiple mutations at once
+	const [isRunningMutation, setIsRunningMutation] = useState<boolean>(false)
+
+	// We pass these progress-tracking refs to the consumer
+	const [startRef, startInView] = useInView()
+	const [endRef, endInView] = useInView()
+
+	/**
+	 * When relevant state changes, we ensure tutorial progress is reflected
+	 * accurately by calling tutorial progress mutations.
+	 */
+	useEffect(() => {
+		/**
+		 * If we don't have the tutorial and collection ids we need,
+		 * or if we're not authenticated,
+		 * or if we already in the middle of a mutation,
+		 * then bail early
+		 */
+		const hasIds = tutorialId && collectionId
+		if (!hasIds || !isAuthenticated || isRunningMutation) {
+			return
+		}
+		/**
+		 * Determine the new progress state, which we'll then update
+		 * - 'complete' (100%) if the end ref is visible
+		 * - 'in_progress' (50%) if the start content ref is no longer visible
+		 * - 'visited' (0%) in all other cases
+		 */
+		let newProgressLabel: TutorialProgressLabel
+		if (endInView) {
+			newProgressLabel = TutorialProgressLabel.complete
+		} else if (!startInView) {
+			newProgressLabel = TutorialProgressLabel.in_progress
+		} else {
+			newProgressLabel = TutorialProgressLabel.visited
+		}
+		/**
+		 * If we have positive progress, or need to record initial progress,
+		 * then we create or update our progress (formatted as percent for the API).
+		 *
+		 * Note that we don't ever downgrade the percent value of progress.
+		 */
+		const existingPercent = progressLabelToPercent(tutorialProgressLabel)
+		const newPercent = progressLabelToPercent(newProgressLabel)
+		const hasPositiveProgress = parseInt(newPercent) > parseInt(existingPercent)
+		const needsInitialProgress = tutorialProgressLabel === null
+		if (needsInitialProgress || hasPositiveProgress) {
```

Comment:
We don't want to make negative progress updates - if the API says progress is `100` percent, even if we've just loaded a view and can't see `endRef` we don't want to downgrade this progress.

We do need to ensure that if the API progress is `null`, we create a new progress record.